network/benches/socket_bench.rs

```diff
@@ -306,5 +307,91 @@ fn socket_bench(c: &mut Criterion) {
     c.bench("socket_send_throughput", bench);
 }
 
-criterion_group!(network_benches, socket_bench);
+/// Concurrently dial into server
+fn bench_client_connection<F, T, S>(
+    b: &mut Bencher,
+    concurrency: u64,
+    transport_func: F,
+    server_addr: NetworkAddress,
+) where
+    F: Fn() -> T,
+    T: Transport<Output = S> + Send + Sync + 'static,
+    S: AsyncRead + AsyncWrite + Unpin + Send + 'static,
+{
+    let mut runtime = Builder::new()
+        .threaded_scheduler()
+        .core_threads(concurrency as usize)
+        .enable_all()
+        .build()
+        .unwrap();
+    b.iter_with_setup(
+        || {
+            let mut futures = vec![];
+            for _ in 0..concurrency {
+                let transport = transport_func();
+                let addr = server_addr.clone();
+                loop {
+                    if let Ok(fut) = transport.dial(addr.clone()) {
+                        futures.push(async move {
+                            match fut.await {
+                                Ok(_socket) => (),
+                                Err(e) => error!("Failed to upgrade {:?}", e),
+                            };
+                        });
+                        break;
+                    }
+                }
+            }
+            futures
+        },
+        |fut| {
+            let mut handles = FuturesUnordered::new();
+            for f in fut {
+                handles.push(runtime.spawn(f));
+            }
+            runtime.block_on(async move { while let Some(Ok(_)) = handles.next().await {} });
+        },
+    );
+}
+
+/// Measure the concurrent connection throughput of the servers, use with socket-bench-server.
```

Comment:
would be great to explain in detail what the strategy is here, an example output and how to read it
also how variosu numbers were chosen (16, 32, 64, 128, 1024)... teach future @zekun000's so they don't have to learn this by navigating through mio's code base

Reply:
it's chosen randomly be me :)